lib/commands/recordscreen.js

```diff
@@ -96,119 +192,58 @@ async function uploadRecordedMedia (localFile, remotePath = null, uploadOptions
-      'recordVideo'
-    ];
-    if (util.hasValue(videoType)) {
-      args.push('--type', videoType);
-    }
-    args.push(localPath);
+  const screenRecorder = new ScreenRecorder(this.opts.device.udid, videoPath, {
+    // TODO: Apply type and quality options
+    videoType,
+    videoQuality,
+    videoFps: videoFps || DEFAULT_FPS,
+    remotePort: this.opts.mjpegServerPort || DEFAULT_MJPEG_SERVER_PORT,
+    usePortForwarding: this.isRealDevice(),
+  });
+  if (!await screenRecorder.terminate()) {
+    log.errorAndThrow('Unable to stop screen recording process');
+  }
+  if (this._recentScreenRecorder) {
+    await this._recentScreenRecorder.cleanup();
+    this._recentScreenRecorder = null;
   }
 
-  // wrap in a manual Promise so we can handle errors in exec operation
-  return await new B(async (resolve, reject) => {
-    let err = null;
-    let timeout = Math.floor(parseFloat(timeLimit) * 1000);
-    if (timeout > MAX_RECORDING_TIME_SEC * 1000 || timeout <= 0) {
-      return reject(new Error(`The timeLimit value must be in range (0, ${MAX_RECORDING_TIME_SEC}] seconds. ` +
-                              `The value of ${timeLimit} has been passed instead.`));
-    }
-    log.debug(`Beginning screen recording with command: '${binaryName} ${args.join(' ')}'` +
-              `Will timeout in ${timeout / 1000} s`);
-    if (this.isRealDevice()) {
-      // xrecord has its owen timer, so we only use this one as a safety precaution
-      // although simctl has no built-in timer and we have to be precise in such case
-      timeout += PROCESS_SHUTDOWN_TIMEOUT_SEC * 1000 * 2;
-    }
-    // do not await here, as the call runs in the background and we check for its product
-    exec(binaryName, args, {timeout, killSignal: 'SIGINT'}).catch((e) => {
-      err = e;
-    });
+  const timeout = Math.floor(parseFloat(timeLimit) * 1000);
+  if (timeout > MAX_RECORDING_TIME_SEC * 1000 || timeout <= 0) {
+    log.errorAndThrow(`The timeLimit value must be in range (0, ${MAX_RECORDING_TIME_SEC}] seconds. ` +
+      `The value of ${timeLimit} has been passed instead.`);
+  }
 
-    // there is the delay time to start recording the screen for real devices, so, wait until it is ready.
-    // the ready condition is
-    //   1. check the movie file is created
-    //   2. check the screen capture has been started
-    //
-    // simctl keeps the file in an internal buffer instead and only creates it when the recording is done.
-    if (this.isRealDevice()) {
-      try {
-        await retryInterval(10, RETRY_PAUSE, async () => {
-          if (err) {
-            return;
-          }
-
-          const {size} = await fs.stat(localPath);
-          if (size <= 32) {
-            throw new Error(`Remote file '${localPath}' found but it is still too small: ${size} bytes`);
-          }
-        });
-      } catch (e) {
-        err = e;
-      }
-    }
+  try {
+    await screenRecorder.start(timeout);
+  } catch (e) {
+    await screenRecorder.terminate();
```

Comment:
these could be in a `finally` block so we dn't have to explicitly throw

Reply:
we only need to terminate the stuff if an exception happens to properly delete the leftovers on the machine. Although the `finally` block would be always executed, which is undesired.